File: src/main/kotlin/sunsetsatellite/vm/sunlite/SLClosure.kt
package sunsetsatellite.vm.sunlite

class SLClosure(val function: SLFunction, val upvalues: Array<SLUpvalue?> = arrayOfNulls(function.upvalueCount)) {

	override fun toString(): String {
		return function.toString()
	}

	fun copy(): SLClosure {
		return SLClosure(function, upvalues.map { it?.copy() }.toTypedArray())
	}
}

File: src/main/csharp/SunliteSharp/Runtime/SLClosure.cs
namespace SunliteSharp.Runtime;

public record SLClosure(SLFunction Function, SLUpvalue?[] Upvalues)
{
    public override string ToString()
    {
        return Function.ToString();
    }

    public SLClosure Copy()
    {
        return this with { Upvalues = Upvalues.Select(u => u?.Copy()).ToArray() };
    }
}